apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl

```diff
@@ -0,0 +1,209 @@
+    next/2
+]).
+
+%% internal exports:
+-export([
+    do_open_shard_v1/2,
+    do_drop_shard_v1/1,
+    do_get_streams_v1/3,
+    do_make_iterator_v1/3,
+    do_next_v1/3
+]).
+
+-export_type([shard_id/0, stream/0, iterator/0, message_id/0]).
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type db() :: emqx_ds:db().
+
+-type shard_id() :: {emqx_ds:db(), atom()}.
+
+%% This record enapsulates the stream entity from the replication
+%% level.
+%%
+%% TODO: currently the stream is hardwired to only support the
+%% internal rocksdb storage. In the future we want to add another
+%% implementations for emqx_ds, so this type has to take this into
+%% account.
+-record(stream, {
+    shard :: emqx_ds_replication_layer:shard_id(),
+    enc :: emqx_ds_storage_layer:stream()
+}).
+
+-opaque stream() :: #stream{}.
+
+-record(iterator, {
+    shard :: emqx_ds_replication_layer:shard_id(),
+    enc :: enqx_ds_storage_layer:iterator()
+}).
+
+-opaque iterator() :: #iterator{}.
+
+-type message_id() :: emqx_ds_storage_layer:message_id().
+
+%%================================================================================
+%% API functions
+%%================================================================================
+
+-spec list_shards(emqx_ds:db()) -> [shard_id()].
+list_shards(DB) ->
+    %% TODO: milestone 5
+    lists:map(
+        fun(Node) ->
+            shard_id(DB, Node)
+        end,
+        list_nodes()
+    ).
+
+-spec open_db(emqx_ds:db(), emqx_ds:create_db_opts()) -> ok | {error, _}.
+open_db(DB, Opts) ->
+    %% TODO: improve error reporting, don't just crash
+    lists:foreach(
+        fun(Node) ->
+            Shard = shard_id(DB, Node),
+            ok = emqx_ds_proto_v1:open_shard(Node, Shard, Opts)
+        end,
+        list_nodes()
+    ).
+
+-spec drop_db(emqx_ds:db()) -> ok | {error, _}.
+drop_db(DB) ->
+    lists:foreach(
+        fun(Node) ->
+            Shard = shard_id(DB, Node),
+            ok = emqx_ds_proto_v1:drop_shard(Node, Shard)
+        end,
+        list_nodes()
+    ).
+
+-spec store_batch(emqx_ds:db(), [emqx_types:message()], emqx_ds:message_store_opts()) ->
+    emqx_ds:store_batch_result().
+store_batch(DB, Msg, Opts) ->
+    %% TODO: Currently we store messages locally.
+    Shard = shard_id(DB, node()),
+    emqx_ds_storage_layer:store_batch(Shard, Msg, Opts).
+
+-spec get_streams(db(), emqx_ds:topic_filter(), emqx_ds:time()) ->
+    [{emqx_ds:stream_rank(), stream()}].
+get_streams(DB, TopicFilter, StartTime) ->
+    Shards = emqx_ds_replication_layer:list_shards(DB),
+    lists:flatmap(
+        fun(Shard) ->
+            Node = node_of_shard(Shard),
+            Streams = emqx_ds_proto_v1:get_streams(Node, Shard, TopicFilter, StartTime),
+            lists:map(
+                fun({RankY, Stream}) ->
+                    RankX = erlang:phash2(Shard, 255),
```

Comment:
Maybe it would be safer to use a larger range, or another hashing method?

In 2 naive tests, I got a collision with 3 elements and one with 2 (one of them is not exactly of the same type as our shard ids here, I know):

```elixir
iex(emqx@172.100.239.2)22> 1..20 |> Enum.map(& "emqx@emqx#{&1}.net") |> Enum.map(& :erlang.phash2(&1, 255)) |> Enum.group_by(& &1) |> Map.filter(fn {_k, vs} -> length(vs) > 1 end)                
%{137 => [137, 137, 137]}
iex(emqx@172.100.239.2)23> 1..20 |> Enum.map(& {:ps, String.to_atom("emqx@emqx#{&1}.net")}) |> Enum.map(& :erlang.phash2(&1, 255)) |> Enum.group_by(& &1) |> Map.filter(fn {_k, vs} -> length(vs) > 1 end)
%{146 => [146, 146]}
```

Reply:
Very good catch. A hash collision here could create an artificial dependency between unrelated shards.

Reply:
Good idea, that would remove any collision possibility of this "birthday paradox".